### ISSUE TITLE
Nvidia Shield - Widget doesn't auto-refresh

### ISSUE THREAD

**Issue by TtuxX** [stefan-niedermann/nextcloud-deck#1436]
Hi there! :)

I hope you are both doing fine and that everything goes well for you.

I am using your great app as a widget on my Nvidia Shield Pro 2019 (to serve as a reminder of the things I need to buy and do haha) and I am having an issue regarding background refresh.

I disabled "battery" optimisation but despite trying different auto-refresh settings inside your app, the displayed widget does not appear to auto-refresh. 
DAVx5 app is for example refreshing fine in background, but not your app.
Only a reboot, or manually opening your app appears to refresh it.

I am using Wolf Launcher, if it has any kind of importance.

I thank you in advance for your kind help and remain at your disposal to provide you with any useful information.

**Comment by stefan-niedermann** [comment]
Hey,

- which widget are you using? Single Card, Upcoming Cards or List?
- Please try switching to the default launcher. It is very well possible that third party launchers are causing issues.
- Are the information in the app itself (not widget) synvhronized correctly?

**Comment by TtuxX** [comment]
Hi !

- I am using List widget
- I have removed the default one, I have to test, however given the next bullet point, I think it is not the launcher (furthermore, Business Calendar widget works fine, Davx5 works fine, and Chronus widgets work fine)
- The information in the app itself are not synchronised correctly, and indeed only synchronise when I manually open the app.

The app version is 1.21.6 (I tried updating to 1.21.8 but this version appears to have some other issues on my device and import seems to be stalled forever on deck 1, so for now 1.21.6 is more stable for me)

**Comment by TtuxX** [comment]
Well after trying for several days I am so wrong and you are terribly right: Wolf Launcher has widget updating issues.
All my sincere apologies for opening this issue.
For anyone encountering the same issue: force close the launcher then open it again. If widget has updated, then you do have an issue with your launcher like I do.

**Comment by desperateCoder** [comment]
LoL... No, buddy! Stuff isn't always obvious. Especially regarding this kind of fragmentation. All good, better we have a look than failing at basic features. But big thanks for the feedback, now we know!

**Comment by TtuxX** [comment]
Well well well... here I am again ! haha This time with some error logs. 
But I am not sure this is related to/a side effect of the refresh issue.

So:
- Widget refresh issue popped-up again after some time the Shield was on
- The widget was not refreshing whereas those from other apps were
- I tried to "configure" the widget in-place, to select another deck column from the same board and save the new settings
- Then, upon clicking on "Add", I repeatedly get the following error:

```
App Version: 1.21.8
App Version Code: 1021008
Server App Version: 1.8.3
App Flavor: fdroid

Files App Version Code: 20220902 (Beta)

---

OS Version: 4.9.141-tegra-gfad87a91fd8a(7825230_3167.5736)
OS API Level: 30
Device: mdarcy
Manufacturer: NVIDIA
Model (and Product): SHIELD Android TV (mdarcy)

---

android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: FilterWidget.id (code 1555 SQLITE_CONSTRAINT_PRIMARYKEY)
	at android.database.sqlite.SQLiteConnection.nativeExecuteForLastInsertedRowId(Native Method)
	at android.database.sqlite.SQLiteConnection.executeForLastInsertedRowId(SQLiteConnection.java:938)
	at android.database.sqlite.SQLiteSession.executeForLastInsertedRowId(SQLiteSession.java:790)
	at android.database.sqlite.SQLiteStatement.executeInsert(SQLiteStatement.java:88)
	at androidx.sqlite.db.framework.FrameworkSQLiteStatement.executeInsert(FrameworkSQLiteStatement.kt:42)
	at androidx.room.EntityInsertionAdapter.insertAndReturnId(EntityInsertionAdapter.kt:102)
	at it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao.widgets.filter.FilterWidgetDao_Impl.insert(FilterWidgetDao_Impl.java:140)
	at it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao.widgets.filter.FilterWidgetDao_Impl.insert(FilterWidgetDao_Impl.java:27)
	at it.niedermann.nextcloud.deck.persistence.sync.adapters.db.DataBaseAdapter.createFilterWidgetDirectly(DataBaseAdapter.java:1203)
	at it.niedermann.nextcloud.deck.persistence.sync.SyncManager.lambda$createFilterWidget$51$it-niedermann-nextcloud-deck-persistence-sync-SyncManager(SyncManager.java:1866)
	at it.niedermann.nextcloud.deck.persistence.sync.SyncManager$$ExternalSyntheticLambda21.run(Unknown Source:6)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:923)
```

**Comment by desperateCoder** [comment]
Welcome back @TtuxX 😅 

From seeing the stack trace above i can tell it isn't related to the refresh issues. I suspect the UI is trying to persist your changes, but instead of updating the existing widget information, it tries to create the same widget again, which gets blocked by the database, since it's id has to be unique. 

We'll have a look at it for sure, but unfortunately it won't fix the refresh problem.

**Comment by TtuxX** [comment]
Ok, than you for this info :)
I am remain at your disposal in case I can be of any help to understand where the refresh issue comes from.

**Comment by TtuxX** [comment]
Some more info that might be useful:
- When I manually launch the Deck app, it syncs and widget gets updated (it seems that the background service it not properly executed)
- Background sync is set to 15 minutes
- "Only sync in WIFI" toogle is disabled
- Battery optimisation is disabled

**Comment by TtuxX** [comment]
Closing this issue as you were right.
With latest ATV Launcher Pro version (Wolf Launcher is a stolen version of it), widget refresh has been working perfectly fine for weeks now. 
I am a happy user, delighted to have my synchonised To-Do on my TV home screen thanks to you and your great Android NC Deck app 🎉

**Comment by stefan-niedermann** [comment]
Thanks for the update!
